### PR TITLE
feat: update response and error handling

### LINKED ISSUE #93: js sdk improve exception ux

### PR DESCRIPTION
**Pull request by pgautier404**
This commit updates response and error handling to bring them into conformance with [error handling standards and practices](https://github.com/momentohq/standards-and-practices/blob/main/docs/client-specifications/error-handling.md).

I'll update the examples and the README files once this is merged and I can release a new version of the SDK.

Fixes #93

### CODE REVIEW

**Comment by cprice404** [comment]
@pgautier404 this looks really nice at first glance!  There's more here than I will be able to take in today.  One thing that would be really helpful is if you could file a second PR that is just a "draft" PR that would show how the example code will change to correspond to these changes.

I'd like to have some combination of @bruuuuuuuce , @eaddingtonwhite , and @poppoerika take a look at the new API and see how it will feel, maybe even pull it down and kick the tires a bit.  But being able to show what the diff of the example code will look like I think will go a long way towards helping folks evaluate UX.

File: integration/simple-cache-client.test.ts [hunk 1]
```diff
@@ -60,87 +67,151 @@ describe('SimpleCacheClient.ts Integration Tests', () => {
     await withCache(momento, cacheName, async () => {
       await momento.set(cacheName, 'key', 'value');
       const res = await momento.get(cacheName, 'key');
-      expect(res.text()).toEqual('value');
+      expect(res instanceof CacheGet.Error).toEqual(false);
+      if (res instanceof CacheGet.Hit) {
+        expect(res.valueString()).toEqual('value');
+      }
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:73]: If it isn't a CacheGet.Hit, the test will pass. Same for the rest of the tests.

Add an expectation.

```suggestion
      expect(res instanceof CacheGet.Hit).toEqual(true)
      expect(res.valueString()).toEqual('value');
```

File: integration/simple-cache-client.test.ts [hunk 2]
```diff
@@ -60,87 +67,151 @@ describe('SimpleCacheClient.ts Integration Tests', () => {
...
+  it('should return NotFoundError if deleting a non-existent cache', async () => {
     const cacheName = v4();
     const momento = new SimpleCacheClient(AUTH_TOKEN, 1111);
-    await expect(momento.deleteCache(cacheName)).rejects.toThrow(NotFoundError);
+    const deleteResponse = await momento.deleteCache(cacheName);
+    expect(deleteResponse instanceof DeleteCache.Error).toEqual(true);
+    if (deleteResponse instanceof DeleteCache.Error) {
+      expect(deleteResponse.errorCode()).toEqual(
+        MomentoErrorCode.NOT_FOUND_ERROR
+      );
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:84]:
```suggestion
    expect(deleteResponse instanceof DeleteCache.Error).toEqual(true);
    expect(deleteResponse.errorCode()).toEqual(
      MomentoErrorCode.NOT_FOUND_ERROR
    );
```

File: integration/simple-cache-client.test.ts [hunk 3]
```diff
@@ -60,87 +67,151 @@ describe('SimpleCacheClient.ts Integration Tests', () => {
...
-      const caches = (await momento.listCaches()).getCaches();
-      const names = caches.map(c => c.getName());
-      expect(names.includes(cacheName)).toBeTruthy();
+      const listResponse = await momento.listCaches();
+      expect(listResponse instanceof ListCaches.Error).toEqual(false);
+      if (listResponse instanceof ListCaches.Success) {
+        const caches = listResponse.getCaches();
+        const names = caches.map(c => c.getName());
+        expect(names.includes(cacheName)).toBeTruthy();
+      }
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:105]:
```suggestion
      expect(listResponse instanceof ListCaches.Error).toEqual(false);
      expect(listResponse instanceof ListCaches.Success).toEqual(true)

      const names = listResponse.getCaches().map(c => c.getName());
      expect(names.includes(cacheName)).toBeTruthy();
```

File: integration/simple-cache-client.test.ts [hunk 4]
```diff
@@ -60,87 +67,151 @@ describe('SimpleCacheClient.ts Integration Tests', () => {
...
+    expect(setResponse instanceof CacheSet.Error).toEqual(false);
+    const getResponse = await momento.get(
+      INTEGRATION_TEST_CACHE_NAME,
+      cacheKey
+    );
+    expect(getResponse instanceof CacheGet.Error).toEqual(false);
+    expect(getResponse instanceof CacheGet.Miss).toEqual(false);
+    if (getResponse instanceof CacheGet.Hit) {
+      expect(getResponse.valueString()).toEqual(cacheValue);
+    }
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:126]:
```suggestion
    expect(getResponse instanceof CacheGet.Hit).toEqual(true);
    expect(getResponse.valueString()).toEqual(cacheValue);
```

And so on.

File: integration/simple-cache-client.test.ts [hunk 5]
```diff
@@ -152,42 +223,97 @@ describe('SimpleCacheClient.ts Integration Tests', () => {
...
+    const deleteResponse = await momento.delete(
+      INTEGRATION_TEST_CACHE_NAME,
+      cacheKey
+    );
+    expect(deleteResponse instanceof CacheDelete.Error).toEqual(false);
+    const missResponse = await momento.get(
+      INTEGRATION_TEST_CACHE_NAME,
+      cacheKey
+    );
+    expect(missResponse instanceof CacheGet.Miss).toEqual(true);
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:272]: This seems redundant with the previous test.

File: src/errors/errors.ts [hunk 6]
```diff
@@ -1,153 +1,201 @@
...
  */
-export class InvalidArgumentError extends ClientSdkError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'InvalidArgumentError';
+export abstract class SdkError extends Error {
+  public readonly errorCode: MomentoErrorCode;
+  public readonly messageWrapper: string;
+  public readonly transportDetails: MomentoErrorTransportDetails;
+  public readonly stack: string | undefined;
```
Comment (schwern) [inline src/errors/errors.ts:58]: Why is this undefined instead of null?

File: src/internal/cache-client.ts [hunk 7]
```diff
@@ -208,36 +213,37 @@ export class CacheClient {
...
+                break;
+              case cache.cache_client.ECacheResult.Hit:
+                resolve(new CacheGet.Hit(resp.cache_body));
+                break;
+              case cache.cache_client.ECacheResult.Invalid:
+                resolve(new CacheGet.Error(new UnknownError(resp.message)));
+                break;
+              case cache.cache_client.ECacheResult.Ok:
+                resolve(new CacheGet.Error(new UnknownError(resp.message)));
+                break;
```
Comment (schwern) [inline src/internal/cache-client.ts:228]: The two cases seem the same. Use a fall through?
```suggestion
              case cache.cache_client.ECacheResult.Invalid:
              case cache.cache_client.ECacheResult.Ok:
                resolve(new CacheGet.Error(new UnknownError(resp.message)));
                break;
```

**Review by schwern — COMMENTED** [review]
This is as far in as I got.

File: src/errors/errors.ts (hunk 6 above)
Comment (cprice404) [reply src/errors/errors.ts:58]: This Is The Way in typescript :)

this is the way

File: integration/simple-cache-client.test.ts (hunk 3 above)
Comment (cprice404) [reply integration/simple-cache-client.test.ts:105]: `getCaches` is only available if the type of the response is `Success`.  Putting the code inside of the instanceof guard statement allows the compiler to smart-cast to `Success` and thus makes the `getCaches()` method available when it would not otherwise be available.

My expectation would be that the version of the code you pasted here would not compile, because the compiler will only be able to infer the type of `listResponse` to be `ListCachesResponse`, and that class does not have a method called `getCaches`.

File: integration/simple-cache-client.test.ts (hunk 1 above)
Comment (cprice404) [reply integration/simple-cache-client.test.ts:73]: I made a similar comment farther down, but I do not believe this would compile because the parent class does not have a method called `valueString`.

File: integration/simple-cache-client.test.ts (hunk 2 above)
Comment (cprice404) [reply integration/simple-cache-client.test.ts:84]: same here.

File: integration/simple-cache-client.test.ts (hunk 1 above)
Comment (pgautier404) [reply integration/simple-cache-client.test.ts:73]: The line before this `if` asserts that the response is not an `.Error` type, so it will fail if an error is encountered. However, since this is a `get` test, it will give a false positive if the result is a miss, so I'll update for that.

File: integration/simple-cache-client.test.ts (hunk 2 above)
Comment (pgautier404) [reply integration/simple-cache-client.test.ts:84]: If I take the second `expect` out of the `if`, I lose the pattern matching/casting to the subclass. I could add an `else` clause to the `if` statement instead, of using a redundant `instanceof` call, but I don't think that really buys me anything.

File: integration/simple-cache-client.test.ts (hunk 3 above)
Comment (pgautier404) [reply integration/simple-cache-client.test.ts:105]: Yes, the `if` is necessary for the cast.

File: integration/simple-cache-client.test.ts (hunk 5 above)
Comment (pgautier404) [reply integration/simple-cache-client.test.ts:272]: Good eye, I deleted it.

File: src/internal/cache-client.ts (hunk 7 above)
Comment (pgautier404) [reply src/internal/cache-client.ts:228]: Sure, will do.

File: integration/simple-cache-client.test.ts (hunk 3 above)
Comment (cprice404) [reply integration/simple-cache-client.test.ts:105]: I do agree with @schwern though that the positive assertion is better than the negative ones.  However unlikely, if we were to ever add more subclasses to the response type, then without the `true` assertion for the `Success` type, the test could be a false positive.

Comment (pgautier404) [reply integration/simple-cache-client.test.ts:105]: I was totally assuming these responses would always be binary success or failure, but I take your point about there possibly being other response types in the future though. I'll fix these up.

Comment (cprice404) [reply integration/simple-cache-client.test.ts:105]: 99.9% chance that your assumption is fine :)  but might as well just make the tests 💯

**Comment by pgautier404** [comment]
> @pgautier404 this looks really nice at first glance! There's more here than I will be able to take in today. One thing that would be really helpful is if you could file a second PR that is just a "draft" PR that would show how the example code will change to correspond to these changes.
> 
> I'd like to have some combination of @bruuuuuuuce , @eaddingtonwhite , and @poppoerika take a look at the new API and see how it will feel, maybe even pull it down and kick the tires a bit. But being able to show what the diff of the example code will look like I think will go a long way towards helping folks evaluate UX.

Good idea: https://github.com/momentohq/client-sdk-javascript/pull/132.

File: integration/simple-cache-client.test.ts [hunk 8]
```diff
@@ -61,13 +61,13 @@ async function withCache(
...
 describe('SimpleCacheClient.ts Integration Tests', () => {
-  it('should create and delete a cache', async () => {
+  it('should create and delete a cache, set and get a value', async () => {
     const cacheName = v4();
     const momento = new SimpleCacheClient(AUTH_TOKEN, 1111);
     await withCache(momento, cacheName, async () => {
       await momento.set(cacheName, 'key', 'value');
       const res = await momento.get(cacheName, 'key');
-      expect(res instanceof CacheGet.Error).toEqual(false);
+      expect(res instanceof CacheGet.Hit).toEqual(true);
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:70]: Note there is [`toBeInstanceOf`](https://jestjs.io/docs/expect#tobeinstanceofclass). More concise, easier to read.

```suggestion
      expect(res).toBeInstanceOf(CacheGet.Hit)
```

File: src/messages/responses/create-signing-key/create-signing-key.ts [hunk 9]
```diff
@@ -34,3 +37,27 @@ export class CreateSigningKeyResponse {
...
+  }
+
+  public errorCode(): MomentoErrorCode {
+    return this._innerException.errorCode;
+  }
+
+  public override toString(): string {
+    return super.toString() + ': ' + this.message();
+  }
+}
```
Comment (schwern) [inline src/messages/responses/create-signing-key/create-signing-key.ts:63]: This appears to be a duplicate of CreateCacheResponse's Error. Can this be DRYed up with a [mixin](https://www.typescriptlang.org/docs/handbook/mixins.html)?

File: src/messages/responses/list-caches/list-caches.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    );
+  }
+
+  public getNextToken() {
+    return this.nextToken;
+  }
+
+  public getCaches() {
+    return this.caches;
+  }
```
Comment (schwern) [inline src/messages/responses/list-caches/list-caches.ts:24]: The Error classes use an public attribute() / private _attribute` pattern. This uses a public getAttribute() / private attribute` pattern. Would it be better to make them consistent?

File: src/messages/responses/list-caches/list-caches.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+export class Success extends ListCachesResponse {
+  private readonly nextToken: string | null;
+  private readonly caches: CacheInfo[];
+  constructor(result?: control.control_client._ListCachesResponse) {
+    super();
+    this.nextToken = result?.next_token || null;
+    this.caches = [];
+    result?.cache.forEach(cache =>
+      this.caches.push(new CacheInfo(cache.cache_name))
+    );
```
Comment (schwern) [inline src/messages/responses/list-caches/list-caches.ts:15]: Looks like a map?

File: src/messages/responses/set/cache-set.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    this.value = value;
+  }
+
+  public valueString(): string {
+    return TEXT_DECODER.decode(this.value);
+  }
+
+  public valueBytes(): Uint8Array {
+    return this.value;
+  }
```
Comment (schwern) [inline src/messages/responses/set/cache-set.ts:21]: Mixin to share with CacheGetResponse?

File: src/messages/responses/create-signing-key/create-signing-key.ts (hunk 9 above)
Comment (pgautier404) [reply src/messages/responses/create-signing-key/create-signing-key.ts:63]: I'll add an issue to look at this as a followup. I looked into mixins for this but spent too much time wrestling with it to justify adding it in this PR.

File: src/messages/responses/list-caches/list-caches.ts (hunk 10 above)
Comment (pgautier404) [reply src/messages/responses/list-caches/list-caches.ts:24]: Yes, good call. I'll update the errors.

File: src/messages/responses/list-caches/list-caches.ts (hunk 11 above)
Comment (pgautier404) [reply src/messages/responses/list-caches/list-caches.ts:15]: Yeah, it does. This was copied and pasted from the old response class. Kinda seems six one half dozen the other to me but if you have a compelling reason to change it I'll do it.

File: integration/simple-cache-client.test.ts (hunk 8 above)
Comment (pgautier404) [reply integration/simple-cache-client.test.ts:70]: Nice, thanks.

File: src/messages/responses/set/cache-set.ts (hunk 12 above)
Comment (pgautier404) [reply src/messages/responses/set/cache-set.ts:21]: See above regarding mixins . . . I'll look into that separately.

File: src/messages/responses/list-caches/list-caches.ts (hunk 11 above)
Comment (schwern) [reply src/messages/responses/list-caches/list-caches.ts]: Takes more time to figure out what's going on. And everywhere else uses map.

File: integration/simple-cache-client.test.ts [hunk 13]
```diff
@@ -19,24 +25,32 @@ const deleteCacheIfExists = async (
...
   cacheName: string
 ) => {
-  try {
-    await momento.deleteCache(cacheName);
-  } catch (e) {
-    if (!(e instanceof NotFoundError)) {
-      throw e;
+  const deleteResponse = await momento.deleteCache(cacheName);
+  if (deleteResponse instanceof DeleteCache.Error) {
+    deleteResponse.toString();
```
Comment (schwern) [inline integration/simple-cache-client.test.ts:30]: This seems to do nothing.

**Review by schwern — APPROVED** [review]

File: src/messages/responses/list-caches/list-caches.ts (hunk 11 above)
Comment (pgautier404) [reply src/messages/responses/list-caches/list-caches.ts]: ok, i changed it for you.

**Review by schwern — APPROVED** [review]

File: integration/simple-cache-client.test.ts [hunk 14]
```diff
@@ -19,24 +25,31 @@ const deleteCacheIfExists = async (
...
   cacheName: string
 ) => {
-  try {
-    await momento.deleteCache(cacheName);
-  } catch (e) {
-    if (!(e instanceof NotFoundError)) {
-      throw e;
+  const deleteResponse = await momento.deleteCache(cacheName);
+  if (deleteResponse instanceof DeleteCache.Error) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (bruuuuuuuce) [inline integration/simple-cache-client.test.ts:30]: why does it think that `deleteResponse.errorCode()` is potentially unsafe?

File: src/errors/error-utils.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    return this._innerException.errorCode();
+  }
+
+  public toString(): string {
+    return this.message();
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function applyMixins(derivedCtor: any, constructors: any[]) {
```
Comment (bruuuuuuuce) [inline src/errors/error-utils.ts:23]: It is not immediately clear what this function is doing, might be good to add a comment about exactly what its purpose is. Also, could this function be typed with `export function applyMixins(derivedCtor: ErrorBody, constructors: ErrorBody[])`? Just looking at where its used it looks like it could potentially be typed this way.

File: src/errors/errors.ts [hunk 16]
```diff
@@ -1,153 +1,209 @@
...
+  // Request rate exceeded the limits for the account
+  LIMIT_EXCEEDED_ERROR = 'LIMIT_EXCEEDED_ERROR',
+  // Request was invalid
+  BAD_REQUEST_ERROR = 'BAD_REQUEST_ERROR',
+  // Client's configured timeout was exceeded
+  TIMEOUT_ERROR = 'TIMEOUT_ERROR',
+  // Server was unable to handle the request
+  SERVER_UNAVAILABLE = 'SERVER_UNAVAILABLE',
+  // A client resource (most likely memory) was exhausted
+  CLIENT_RESOURCE_EXHAUSTED = 'CLIENT_RESOURCE_EXHAUSTED',
```
Comment (bruuuuuuuce) [inline src/errors/errors.ts:27]: What is the difference between `CLIENT_RESOURCE_EXHAUSTED` and `LIMIT_EXCEEDED_ERROR`?

File: src/errors/errors.ts [hunk 17]
```diff
@@ -1,153 +1,209 @@
...
- * Error when calls are throttled due to request limit rate
+ * Insufficient permissions to perform an operation on Cache Service
  */
-export class LimitExceededError extends CacheServiceError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'LimitExceededError';
-  }
+export class PermissionError extends SdkError {
+  _errorCode = MomentoErrorCode.PERMISSION_ERROR;
```
Comment (bruuuuuuuce) [inline src/errors/errors.ts:171]: Just curious, do we currently ever throw this error? I have never seen it before

File: src/internal/cache-client.ts [hunk 18]
```diff
@@ -156,11 +156,12 @@ export class CacheClient {
...
     cacheName: string,
     key: Uint8Array
-  ): Promise<DeleteResponse> {
+  ): Promise<CacheDelete.Response> {
     const request = new cache.cache_client._DeleteRequest({
       cache_key: key,
     });
     const metadata = this.createMetadata(cacheName);
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
     return await new Promise((resolve, reject) => {
```
Comment (bruuuuuuuce) [inline src/internal/cache-client.ts:165]: nit: can remove `reject` and then you don't need to disable any eslint rules here

File: src/internal/cache-client.ts [hunk 19]
```diff
@@ -182,23 +183,24 @@ export class CacheClient {
...
     key: Uint8Array
-  ): Promise<GetResponse> {
+  ): Promise<CacheGet.Response> {
     const request = new cache.cache_client._GetRequest({
       cache_key: key,
     });
     const metadata = this.createMetadata(cacheName);
 
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
     return await new Promise((resolve, reject) => {
```
Comment (bruuuuuuuce) [inline src/internal/cache-client.ts:204]: ditto

File: src/internal/cache-client.ts [hunk 20]
```diff
@@ -208,36 +210,35 @@ export class CacheClient {
...
+            switch (resp.result) {
+              case cache.cache_client.ECacheResult.Miss:
+                resolve(new CacheGet.Miss());
+                break;
+              case cache.cache_client.ECacheResult.Hit:
+                resolve(new CacheGet.Hit(resp.cache_body));
+                break;
+              case cache.cache_client.ECacheResult.Invalid:
+              case cache.cache_client.ECacheResult.Ok:
+                resolve(new CacheGet.Error(new UnknownError(resp.message)));
```
Comment (bruuuuuuuce) [inline src/internal/cache-client.ts:222]: this reads a little weird to me. If the result is `OK` we return an error? I am assuming this is because get should only return a hit or a miss?

**Review by bruuuuuuuce — COMMENTED** [review]
A few nits/questions but overall looks great!

File: integration/simple-cache-client.test.ts [hunk 21]
```diff
@@ -19,24 +25,31 @@ const deleteCacheIfExists = async (
...
+    if (deleteResponse.errorCode() !== MomentoErrorCode.NOT_FOUND_ERROR) {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (eaddingtonwhite) [inline integration/simple-cache-client.test.ts:32]: why do we need this eslint disabled line on 32 here but not on 43  or 51?

Comment (eaddingtonwhite) [reply integration/simple-cache-client.test.ts:32]: O I guess this is test but just want to double check make sure nothing weird happening here vs other API calls

File: src/errors/error-utils.ts (hunk 15 above)
Comment (eaddingtonwhite) [reply src/errors/error-utils.ts:23]: Agree maybe also just some JS Doc comments on what its doing and why we need this just for another dev coming in and looking at this?

File: src/errors/errors.ts [hunk 22]
```diff
@@ -1,153 +1,209 @@
...
  */
-export class AuthenticationError extends CacheServiceError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'AuthenticationError';
-  }
+export class LimitExceededError extends SdkError {
+  _errorCode = MomentoErrorCode.LIMIT_EXCEEDED_ERROR;
+  _messageWrapper =
+    'Request rate exceeded the limits for this account.  To resolve this error, reduce your request rate, or contact us at support@momentohq.com to request a limit increase';
```
Comment (eaddingtonwhite) [inline src/errors/errors.ts:154]: Limit errors can be for bandwidth and object size as well. Maybe reword this message a bit to be more generic vs just 'rate'?

File: src/internal/control-client.ts [hunk 23]
```diff
@@ -58,75 +54,85 @@ export class ControlClient {
...
-  public async deleteCache(name: string): Promise<DeleteCacheResponse> {
+  public async deleteCache(name: string): Promise<DeleteCache.Response> {
+    try {
+      this.validateCacheName(name);
+    } catch (err) {
+      if (err instanceof SdkError) {
+        return new DeleteCache.Error(err);
+      } else if (err instanceof Error) {
+        return new DeleteCache.Error(new UnknownError(err.message));
+      }
```
Comment (eaddingtonwhite) [inline src/internal/control-client.ts:100]: seen this little catch and wrap or not logic a bit now. Could maybe dry up with a basic utility so could do something like:
```js
    } catch (err) {
      return wrapErrorIfSDKError(DeleteCache.Error, err)
    }
```

File: integration/simple-cache-client.test.ts (hunk 14 above)
Comment (pgautier404) [reply integration/simple-cache-client.test.ts:30]: I honestly don't know . . . let me know if you'd like to look into it with me.

File: src/errors/error-utils.ts (hunk 15 above)
Comment (pgautier404) [reply src/errors/error-utils.ts:23]: Fair point. I'll add some explanation.

File: src/messages/responses/list-caches/list-caches.ts (hunk 11 above)
Comment (cprice404) [reply src/messages/responses/list-caches/list-caches.ts]: +1 to map for this, thanks for changing it

File: src/errors/error-utils.ts (hunk 15 above)
Comment (pgautier404) [reply src/errors/error-utils.ts:23]: > Also, could this function be typed with `export function applyMixins(derivedCtor: ErrorBody, constructors: ErrorBody[])`? Just looking at where its used it looks like it could potentially be typed this way.

The class passed into `derivedCtor` is not of type `ErrorBody` and the `constructors` argument could be passed anything. This is intended to be a general purpose function for using [mixing](https://www.typescriptlang.org/docs/handbook/mixins.html#alternative-pattern) that we happen to be using for a single specific purpose for now.

File: src/errors/errors.ts (hunk 16 above)
Comment (pgautier404) [reply src/errors/errors.ts:27]: The former is something like your laptop running out of memory. The latter is something like you've exceeded your request rate limit.

Comment (cprice404) [reply src/errors/errors.ts:27]: client resources are things like memory on the client side

File: src/internal/cache-client.ts [hunk 24]
```diff
@@ -119,13 +118,14 @@ export class CacheClient {
...
     ttl: number
-  ): Promise<SetResponse> {
+  ): Promise<CacheSet.Response> {
     const request = new cache.cache_client._SetRequest({
       cache_body: value,
       cache_key: key,
       ttl_milliseconds: ttl * 1000,
     });
     const metadata = this.createMetadata(cacheName);
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (cprice404) [inline src/internal/cache-client.ts:128]: matt's comment below applies here too i think

File: src/internal/cache-client.ts (hunk 20 above)
Comment (cprice404) [reply src/internal/cache-client.ts:222]: yeah the wire format is kind of weird on this one, but i think it would be a bug for the server to return anything but Hit here.  this is a little different from the .net sdk though, which does `if Miss ... else return Hit`.

this one feels probably more correct to me though.

File: integration/simple-cache-client.test.ts (hunk 14 above)
Comment (cprice404) [reply integration/simple-cache-client.test.ts:30]: I would like to look into this, I can find some time today.  Not necessarily a blocker for the PR though.

**Comment by cprice404** [comment]
@eaddingtonwhite is there anything here you think is a blocker or should we fast-follow the rest?

File: src/errors/errors.ts (hunk 17 above)
Comment (pgautier404) [reply src/errors/errors.ts:171]: I imagine so. There is a corresponding gRPC status that gets mapped to this error type.

**Review by eaddingtonwhite — APPROVED** [review]
Great work on this @pgautier404!! 🏆  I think GTG can merge with this as is now and then fast follow on some of the tactical feedback still open in here. Overall looks great very excited to get these changes out!

File: src/errors/errors.ts (hunk 22 above)
Comment (schwern) [reply src/errors/errors.ts:154]: @eaddingtonwhite This error message is copied from the [Standards and Practices Error Handling doc](https://github.com/momentohq/standards-and-practices/blob/main/docs/client-specifications/error-handling.md) so it should be updated there as well.

File: src/internal/cache-client.ts (hunk 18 above)
Comment (pgautier404) [reply src/internal/cache-client.ts:165]: Nice, thanks.
